sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/manager/DownloadManagerReleaseDownloader.java

```diff
@@ -0,0 +1,253 @@
+        Uri downloadUrl = mReleaseDetails.getDownloadUrl();
+        AppCenterLog.debug(LOG_TAG, "Start downloading new release from " + downloadUrl);
+        DownloadManager downloadManager = getDownloadManager();
+        DownloadManager.Request request = new DownloadManager.Request(downloadUrl);
+
+        /* Hide mandatory download to prevent canceling via notification cancel or download UI delete. */
+        if (mReleaseDetails.isMandatoryUpdate()) {
+            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN);
+            request.setVisibleInDownloadsUi(false);
+        }
+        long enqueueTime = System.currentTimeMillis();
+        long downloadId = downloadManager.enqueue(request);
+
+        /* Check for if state changed and task not canceled in time. */
+        if (mRequestTask == task) {
+
+            /* Delete previous download. */
+            long previousDownloadId = SharedPreferencesManager.getLong(PREFERENCE_KEY_DOWNLOAD_ID, INVALID_DOWNLOAD_IDENTIFIER);
+            if (previousDownloadId != INVALID_DOWNLOAD_IDENTIFIER) {
+                AppCenterLog.debug(LOG_TAG, "Delete previous download id=" + previousDownloadId);
+                downloadManager.remove(previousDownloadId);
+            }
+
+            /* Store new download identifier. */
+            setDownloadId(downloadId);
+            mListener.onStart(enqueueTime);
+
+            /* Start monitoring progress for mandatory update. */
+            if (mReleaseDetails.isMandatoryUpdate()) {
+                update();
+            }
+        } else {
+
+            /* State changed quickly, cancel download. */
+            AppCenterLog.debug(LOG_TAG, "State changed while downloading, cancel id=" + downloadId);
+            downloadManager.remove(downloadId);
+        }
+    }
+
+    @WorkerThread
+    synchronized void onUpdate() {
+
+        /* Query download manager. */
+        DownloadManager downloadManager = getDownloadManager();
+        try {
+            Cursor cursor = downloadManager.query(new DownloadManager.Query().setFilterById(mDownloadId));
+            if (cursor == null) {
+                throw new NoSuchElementException();
+            }
+            try {
+                if (!cursor.moveToFirst()) {
+                    throw new NoSuchElementException();
+                }
+                int status = cursor.getInt(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_STATUS));
+                if (status == DownloadManager.STATUS_FAILED) {
+                    throw new IllegalStateException();
+                }
+                if (status != DownloadManager.STATUS_SUCCESSFUL) {
+                    long totalSize = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_TOTAL_SIZE_BYTES));
+                    long currentSize = cursor.getLong(cursor.getColumnIndexOrThrow(DownloadManager.COLUMN_BYTES_DOWNLOADED_SO_FAR));
+                    if (mListener.onProgress(currentSize, totalSize)) {
```

Comment:
In develop the CheckDownloadTask was executed in background. Thus before updating UI, we were checking for state change before proceeding (in updateProgressDialog).

Here canceling the task won't cancel doInBackground, which won't cancel this code either. We need a state check and synchronize with Distribute so that it does not change (by running some of the code there to share the lock like before).

Reply:
Disagree here.

On state (release details change) Distribute recreates downloader and listener. In this particular case `hideProgressDialog` will be called (sync is fixed there), so we are good here.